### PR TITLE
feat: update the docker build workflow based on updates from celestia

### PR DESCRIPTION
**Pull request by MSevey**
## Overview

This PR brings in the updated docker build CI from celestia. 
It also adds a simple docker file so that this repo can actively test the workflow like it does in celestia. 
This branch was testing and validated with https://github.com/rollkit/local-da/pull/21

## Summary by CodeRabbit

- **New Features**
  - Introduced a new workflow for testing the Dockerfile pipeline.
  - Added a Dockerfile for building a containerized Go application.
  - Added a Go module and a simple Go application for testing Docker actions.

- **Improvements**
  - Updated environment variable handling and build conditions in the Dockerfile pipeline to enhance build logic and registry configuration.

### CODE REVIEW

File: .github/workflows/reusable_dockerfile_pipeline.yml [hunk 1]
```diff
@@ -91,37 +180,76 @@ jobs:
...
+          #   registry-url: docker.io
+          #   registry-owner: rollkitorg
+          #   run-on-pr: "false"
+          - name: GHCR
+            user-secret: ${{ github.repository_owner }}
+            token-secret: GITHUB_TOKEN
+            registry-url: ghcr.io
+            registry-owner: ${{ needs.prepare-env.outputs.repo_owner }}
+            run-on-pr: "true"
+          # - name: ScaleWay
```
Comment (MSevey) [inline .github/workflows/reusable_dockerfile_pipeline.yml:214]: Leaving commented for now so we know the requirements if we want it.

**Review by Manav-Aggarwal — APPROVED** [review]

**Review by tuxcanfly — APPROVED** [review]
OK.

File: docker-action-test/go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module github.com/celestiaorg/.celestia-github/docker-action-test
+
+go 1.21
```
Comment (yarikbratashchuk) [inline docker-action-test/go.mod:3]: why not latest?

**Review by yarikbratashchuk — APPROVED** [review]

**Comment by yarikbratashchuk** [comment]
haha, too late :)

**Pull request merged**